src/facade/dragonfly_connection.cc

```diff
@@ -1367,4 +1375,45 @@ void Connection::DecreaseStatsOnClose() {
   --stats_->num_conns;
 }
 
+Connection::WeakRef::WeakRef(std::shared_ptr<Connection> ptr, QueueBackpressure* backpressure,
+                             unsigned thread, uint32_t client_id)
+    : ptr_{ptr}, backpressure_{backpressure}, thread_{thread}, client_id_{client_id} {
+}
+
+unsigned Connection::WeakRef::Thread() const {
+  return thread_;
+}
+
+Connection* Connection::WeakRef::Get() const {
+  DCHECK_EQ(ProactorBase::me()->GetPoolIndex(), int(thread_));
+  return ptr_.lock().get();
```

Comment:
What prevents the shared ptr from going dying after `Get()` finishes?
I mean, let's say `ptr_` indeed (weakly) points to some `S` object, and `lock()` succeeded, thus `get()` returns a real pointer. Why can't `S` then be destroyed (because the last owner kills it), thus the returned value is a dangling pointer?
(Please add such documentation explaining it)

Reply:
Oh, I guess it's because it's the same thread, as CHECKed above?